products/views.py

```diff
@@ -1,3 +1,108 @@
-from django.shortcuts import render
+from django.http     import JsonResponse
+from django.http.response import Http404
+from django.views    import View
 
-# Create your views here.
+from products.models import *
+
+class CategoryView(View) :
+  def get(self, request) :
+
+    category_list = [{
+      'id'             : gender_category.id,
+      'name'           : gender_category.name,
+      'main_category'  : [{
+        'id'   : main_category.id,
+        'name' : main_category.name,
+        'sub_category' : [{
+          'id'   : sub_category.id,
+          'name' : sub_category.name
+        } for sub_category in main_category.subcategory_set.all()]
+      } for main_category in gender_category.maincategory_set.all()]
+    } for gender_category in GenderCategory.objects.all().order_by('id')]
+
+    return JsonResponse({'category_list':category_list}, status=200)
+
+class ProductView(View) :
+  def get(self, request) :
+
+    order_request = request.GET.get('sort')
+    
+    SORT_PREFIX = {
+      'recent'    : '-created_at',
+      'old'       : 'created_at',
+      'ascPrice'  : 'price',
+      'descPrice' : '-price'
+    }
+
+    FILTER_PREFIX = {
+      'gender'        : 'sub_category__main_category__gender_category__in',
+      'main_category' : 'sub_category__main_category__in',
+      'sub_category'  : 'sub_category__in',
+      'color'         : 'color__in',
+      'size'          : 'productsize__size__in',
+      'collection'    : 'collection__in',
+      'conscious'     : 'is_conscious__in'
+    }
+
+    filter_set = {
+      FILTER_PREFIX.get(key) : value for (key, value) in dict(request.GET).items() if FILTER_PREFIX.get(key)
+    }
+
+    try :
+      products = Product.objects.filter(**filter_set).order_by(SORT_PREFIX[order_request])
+
+      products_list = [{
+        'id'            : product.id,
+        'name'          : product.name,
+        'price'         : format(int(product.price), ','),
+        'is_new'        : product.is_new,
+        'is_conscious'  : product.is_conscious,
+        'collection_id' : product.collection_id,
+        'color'         : [i.color for i in products.filter(name=product.name)],
```

Comment:
이름이 같은 모든 상품들의 컬러를 반환해줘야 하는 이유가 있나요?
```suggestion
        'colors'         : [product.color for product in products.filter(name=product.name)],
```

Reply:
h&m 사이트를 보면 아래 사진과 같이 같은 제품의 색을 표시합니다. 
연관제품 테이블을 생성하지 않아서 일단은 이름이 같은제품으로 필터링 했습니다.
<img width="520" alt="스크린샷 2021-10-12 오전 9 59 09" src="https://user-images.githubusercontent.com/86424094/136872909-534b6ba4-8d84-4301-957b-acb7024d9550.png">